src/Microsoft.Net.Http.Headers/NameValueHeaderValue.cs

```diff
@@ -0,0 +1,347 @@
+
+            if (string.IsNullOrEmpty(_value))
+            {
+                return string.IsNullOrEmpty(other._value);
+            }
+
+            if (_value[0] == '"')
+            {
+                // We have a quoted string, so we need to do case-sensitive comparison.
+                return (string.CompareOrdinal(_value, other._value) == 0);
+            }
+            else
+            {
+                return (string.Compare(_value, other._value, StringComparison.OrdinalIgnoreCase) == 0);
+            }
+        }
+
+        public static NameValueHeaderValue Parse(string input)
+        {
+            var index = 0;
+            return SingleValueParser.ParseValue(input, ref index);
+        }
+
+        public static bool TryParse(string input, out NameValueHeaderValue parsedValue)
+        {
+            var index = 0;
+            return SingleValueParser.TryParseValue(input, ref index, out parsedValue);
+        }
+
+        public static IList<NameValueHeaderValue> ParseList(IList<string> input)
+        {
+            return MultipleValueParser.ParseValues(input);
+        }
+
+        public static bool TryParseList(IList<string> input, out IList<NameValueHeaderValue> parsedValues)
+        {
+            return MultipleValueParser.TryParseValues(input, out parsedValues);
+        }
+
+        public override string ToString()
+        {
+            if (!string.IsNullOrEmpty(_value))
+            {
+                return _name + "=" + _value;
+            }
+            return _name;
+        }
+
+        internal static void ToString(ICollection<NameValueHeaderValue> values, char separator, bool leadingSeparator,
+            StringBuilder destination)
+        {
+            Contract.Assert(destination != null);
+
+            if ((values == null) || (values.Count == 0))
+            {
+                return;
+            }
+
+            foreach (var value in values)
+            {
+                if (leadingSeparator || (destination.Length > 0))
+                {
+                    destination.Append(separator);
+                    destination.Append(' ');
+                }
+                destination.Append(value.ToString());
+            }
+        }
+
+        internal static string ToString(ICollection<NameValueHeaderValue> values, char separator, bool leadingSeparator)
+        {
+            if ((values == null) || (values.Count == 0))
+            {
+                return null;
+            }
+
+            var sb = new StringBuilder();
+
+            ToString(values, separator, leadingSeparator, sb);
+
+            return sb.ToString();
+        }
+
+        internal static int GetHashCode(ICollection<NameValueHeaderValue> values)
+        {
+            if ((values == null) || (values.Count == 0))
+            {
+                return 0;
+            }
+
+            var result = 0;
+            foreach (var value in values)
+            {
+                result = result ^ value.GetHashCode();
```

Comment:
Not a good hash function - consider doing what Tuple does instead when combining since order matters.